internal/encryption/keyprovider/openbao/client.go

```diff
@@ -0,0 +1,85 @@
+package openbao
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	openbao "github.com/openbao/openbao/api"
+)
+
+type client interface {
+	WriteWithContext(ctx context.Context, path string, data map[string]interface{}) (*openbao.Secret, error)
+}
+
+// service implements missing utility functions from openbao/api such as routing and serialization.
+type service struct {
+	c client
+}
+
+type dataKey struct {
+	Plaintext  []byte
+	Ciphertext []byte
+}
+
+func (s service) generateDataKey(ctx context.Context, keyName string, bitSize int) (dataKey, error) {
+	path := fmt.Sprintf("/transit/datakey/plaintext/%s", keyName)
+
+	secret, err := s.c.WriteWithContext(ctx, path, map[string]interface{}{
+		"bits": bitSize,
+	})
+	if err != nil {
+		return dataKey{}, fmt.Errorf("error sending datakey request to OpenBao: %w", err)
+	}
+
+	key := dataKey{}
+
+	key.Ciphertext, err = retrieveCiphertext(secret)
+	if err != nil {
+		return dataKey{}, err
+	}
+
+	key.Plaintext, err = retrievePlaintext(secret)
+	if err != nil {
+		return dataKey{}, err
+	}
+
+	return key, nil
+}
+
+func (s service) decryptData(ctx context.Context, keyName string, ciphertext []byte) ([]byte, error) {
+	path := fmt.Sprintf("/transit/decrypt/%s", keyName)
```

Comment:
This may need urlencoding though. :)

Reply:
I believe it's the same as `generateDataKey`. `openbao/api`'s `WriteWithContext` method uses `url.URL`'s `Path` field to store the path and it then becomes encoded in `url.RequestURI`. Here is `openbao/api`'s [NewRequest function](https://github.com/openbao/openbao/blob/main/api/client.go#L1193).

Reply:
@janosdebugs thanks for the additional explanations! With internal escape, it could be the valid URL, but not the one we intend it to be. So I agree that's a problem for certain scenarios so I added escaping here: [6bc13d5](https://github.com/opentofu/opentofu/pull/1436/commits/6bc13d5a9efcf0a884e677f1fc62e325f0ffda5d)